### PR TITLE
_copy_temp() to replace Base.copy() usage.

### PR DESCRIPTION
**Pull request by AstitvaAggarwal**
Second PR - Avoid type piracy and define a custom function to copy outputs of supported types. Refer #517 .

### CODE REVIEW

**Comment by AstitvaAggarwal** [comment]
@willtebbutt I'll start small with this function as `Base.copy()` is used in multiple places. If it works for all test cases in `prepare_pullback_cache()` and `value_and_pullback()` . I will switch it in all the rrules, src files.

**Comment by AstitvaAggarwal** [comment]
I think the test failures are unrelated, they even spring up on PR #531 . All my tests pass locally.

**Comment by willtebbutt** [comment]
Once https://github.com/compintell/Mooncake.jl/pull/531 is merged, you should just need to merge the changes from that PR into this one, and CI should start passing again.

**Comment by willtebbutt** [comment]
If you sync this branch with main, it should clear up any errors caused by the issue with AllocCheck.

**Comment by AstitvaAggarwal** [comment]
Although all tests pass here (as expected) I'm still working on the edge case PR for `Base.deepcopy()` so once that is finalized i will add that logic in as well, for now this code should pass all the given test cases. Also the number of commits in between kinda helped me figure out the bug in `deepcopy` 😂.

**Comment by yebai** [comment]
Do we still miss anything here?

**Comment by AstitvaAggarwal** [comment]
@yebai No not really, just waiting for @willtebbutt to take a look.

File: src/interface.jl [hunk 1]
```diff
@@ -238,6 +238,63 @@ end
 
 const _BuiltinArrays = @static VERSION >= v"1.11" ? Union{Array,Memory} : Array
 
+# explicit for svec
+function _copy_temp(x::P) where {P<:SimpleVector}
+    return Core.svec([map(_copy_temp, x_sub) for x_sub in x]...)
+end
```
Comment (willtebbutt) [inline src/interface.jl]:
```suggestion
_copy_temp(x::SimpleVector) = Core.svec([map(_copy_temp, x_sub) for x_sub in x]...)
```
It doesn't look like the parameter `P` is being used in the method body.

File: src/interface.jl [hunk 2]
```diff
@@ -238,6 +238,63 @@ end
...
+    @inbounds for i in eachindex(temp)
+        isassigned(x, i) && (temp[i] = _copy_temp(x[i]))
+    end
+    return temp
+end
+
+# Tuple, NamedTuple
+function _copy_temp(x::P) where {P<:Union{Tuple,NamedTuple}}
+    return map(_copy_temp, x)
+end
```
Comment (willtebbutt) [inline src/interface.jl]:
```suggestion
_copy_temp(x::Union{Tuple,NamedTuple}) = map(_copy_temp, x)
```

File: test/interface.jl [hunk 3]
```diff
@@ -175,5 +175,61 @@ end
                 @test isa(err, Mooncake.ValueAndPullbackReturnTypeError)
             end
         end
+
+        # Test for new copy function
+        @testset for i in eachindex(additional_test_set)
```
Comment (willtebbutt) [inline test/interface.jl]: rather than having a comment, it might be better to include a string in the testset. Something like
```suggestion
        @testset "_copy_temp, $(typeof(i))" for i in eachindex(additional_test_set)
```

Comment (willtebbutt) [reply test/interface.jl]: Oh, wait, sorry. Why are we looping over the indices of `additional_test_set`, rather than just its items?

File: test/interface.jl [hunk 4]
```diff
@@ -175,5 +175,61 @@ end
...
+            original = additional_test_set[i][2]
+            try
+                if isnothing(Mooncake.__exclude_unsupported_output(original))
+                    test_copy = Mooncake._copy_temp(original)
+
+                    function comparisons(
```
Comment (willtebbutt) [inline test/interface.jl]: Does `Mooncake.TestUtils.has_equal_data` not give us the correct definition of equality here?

File: test/interface.jl (hunk 3 above)
Comment (AstitvaAggarwal) [reply test/interface.jl]: sorry, using `eachindex` is almost a habit for me now lol.

File: test/interface.jl (hunk 4 above)
Comment (AstitvaAggarwal) [reply test/interface.jl]: ooh, i havent checked that out, will do. thanks

File: src/interface.jl [hunk 5]
```diff
@@ -238,6 +238,63 @@ end
...
+# mutable composite types, bitstype
+function _copy_temp(x::P) where {P}
+    isbitstype(P) && return x
+    nf = nfields(P)
+
+    if ismutable(x)
+        temp = ccall(:jl_new_struct_uninit, Any, (Any,), P)
+        for x_sub in 1:nf
+            if isdefined(x, x_sub)
+                ccall(
```
Comment (willtebbutt) [inline src/interface.jl:265]: Possibly I'm missing something obvious, but do we have to use this `ccall`? Could we not call `setfield!` instead?

Comment (willtebbutt) [reply src/interface.jl:265]: Oh, is it that `setfield!` not work if a particular field of a `mutable struct` is declared `const`?

Comment (AstitvaAggarwal) [reply src/interface.jl:265]: yes that is the reason.

**Comment by willtebbutt** [comment]
I think the only thing left to do here is to sort the name of the function out. As you say, something other than `_copy_temp` is probably needed.

What do you think about `_copy_output`, since we only ever use it to copy the output of a function? It's not an amazing name, but we can always change it later since we won't be making it part of the public interface.

**Comment by AstitvaAggarwal** [comment]
yeah it sounds okay.

**Review by willtebbutt — APPROVED** [review]
LGTM. Thanks for this!
